Allow overriding asset size limit on parent group level

Martchus opened the pull request:
See https://progress.opensuse.org/issues/56453

---

The fact that we have job groups not within parents makes describing the logic a little bit complicated:

1. If an asset belongs to a job group which is part of a parent group the limit defined on job group level is ignored. Instead, the asset is only accounted to the parent group if it fits in. If it doesn't fit in it doesn't matter if it would have fit into the job group because that limit is ignored.
2. If an asset belongs to a job group which is not part of a parent group the size limit of the job group is used as usual.
3. An asset is only accounted to one parent group or job group at the same time. If an asset is used across groups it will be accounted to the largest parent or job group with enough space according to (1) and (2).

Point 1. means that the statistic for "exclusively kept assets" on job group level would always be zero for job groups within parents. So we need to hide that statistic in that case and instead have it on parent level.

okurz commented:
I like your approach that you try to grasp the general concept first and put it into the description before going into (too much) details. How about – as you already did at least once before – you update the documentation first according to what you think this would imply?

It feels like it this is another point in time where we should think about ditching the concept of parent groups again and instead foresee that every job group can have [0,N] children. WDYT?

okurz commented on docs/UsersGuide.asciidoc:
```diff
@@ -957,20 +957,26 @@ openQA frequently checks whether assets need to be removed according to
 To find out whether an asset should be removed, openQA determines by which
-job groups the asset is used. If at least one job within a certain job group
-is using an asset, the asset is considered to be used by that group.
+groups the asset is used. If at least one job within a certain job group is
```
a *parent job group* is still a *job group* so I would keep this notion. A *group* as word standing on its own is a bit ambiguous. And one can use a specification *child job group* if a *job group* is within a *parent job group*. That would make a *job group* without *parent job group* an *orphan job group* – or *parentless* as you called them further below. But I think the word is simply orphan ;). Did we have this discussion already some months/years ago? ;)

coolo replied:
> a _parent job group_ is still a _job group_ 

No, it's not. It's a group of job groups not a group of jobs. Perhaps we better use 'job group parent' instead to make it clear.

okurz commented on docs/UsersGuide.asciidoc:
```diff
@@ -957,20 +957,26 @@ openQA frequently checks whether assets need to be removed according to
-So an asset can be accounted to multiple groups. The assets table which is
-accessible via the admin menu shows these groups for each asset and also
-the latest job.
+So an asset can belong to multiple job or parent groups. The assets
```
here I would keep the generic *… to multiple job groups*.

okurz commented on docs/UsersGuide.asciidoc:
```diff
@@ -957,20 +957,26 @@ openQA frequently checks whether assets need to be removed according to
-If the size limit for assets by a certain group is exceeded, openQA will
-remove assets accounted to that group:
+While an asset might belong to multiple groups it is only *accounted*
+to the largest group which has still enough room to hold that asset. So
```
"largest group" can be ambiguous, e.g. a job group with most child job groups or a job group with "most tests". Better be explicit about the size quota. And I am not sure about the meaning here. Is it that an asset is assigned to the job group that has the highest amount of free quota or the job group with the highest total quota value?

coolo replied:
This is explaining the algorithm and that can only go wrong - I just wouldn't describe that detail at all.

Martchus replied:
> "largest group" can be ambiguous, e.g. a job group with most child job groups or a job group with "most tests".

None of those options. "largest group" means the group with the highest size limit for assets.

> Is it that an asset is assigned to the job group that has the highest amount of free quota or the job group with the highest total quota value?

As I said, "largest group" refers to "highest total quote" but there's of course the restriction "which has still enough room to hold that asset" so the "free quota" is obviously not ignored, too. Likely @coolo is right an we shouldn't describe the algorithm in that detail.

okurz commented on docs/UsersGuide.asciidoc:
```diff
@@ -989,9 +995,14 @@ in the operators menu and select a group. The size limit for assets can be
+job group is big enough to hold the asset it might not fit into the parent
+group and might therefore be removed.
+
+The default size limit for groups can be adjusted in the +default_group_limits+
```
s/groups/job groups/

okurz commented on etc/openqa/openqa.ini:
```diff
@@ -117,7 +117,8 @@ blacklist = job_grab job_done
 [default_group_limits]
-#asset_size_limit = 100
+#asset_size_limit = 100 # enforced on job group level for job groups without parent
+#parent_asset_size_limit = 500 # enforced on parent group level
```
Do we need two different default levels? I think we should be ok to have just the default as in before

coolo replied:
I agree. For migration I would set the job group parent's limit to the sum of the job group's limits  though.

okurz commented on templates/admin/group/group_property_editor.html.ep:
```diff
@@ -26,13 +26,13 @@
                 <div class="form-group row">
                     <label for="editor-size-limit" class="col-sm-2 control-label" data-toggle="tooltip" title="Size limit for assets">Size limit for assets</label>
-                    <div class="col-sm-10">
+                    <div class="col-sm-10"><!-- FIXME: if there's a parent: add a remark if this limit is overriden on parent level and not used at all until the group is moved outside of the parent -->
```
at best also set this part to "disabled" or greyed-out

coolo commented on docs/UsersGuide.asciidoc:
```diff
@@ -989,9 +995,14 @@ in the operators menu and select a group. The size limit for assets can be
-is set explicitely. The default size limit for groups can be adjusted in
-the +default_group_limits+ section of the openQA config file.
+If a job group is part of a parent group the size limit is set on parent
+group level and the limit set on job group level is ignored. So even if the
```
this should then also be removed - as there is no limit on job groups in parents worth it.

Martchus commented:
I've played around with this on my local instance and the algorithm actually does what it is supposed to do. At least with the few assets I have registered here it makes sense. I adapted the admin page to support parent groups, too. The "size of exclusively kept assets" is also already computed for parent groups.

So I now need to tweak remaining details and of course add some unit tests. I guess I can also test with OSD data locally if I temporarily mock the function which checks whether assets are actually present in the file system.

Martchus commented:
I extended some of the unit tests. There were 2 fixtrue job groups anyways so I moved one into a parent (for those specific tests) and kept one stand alone. This should cover the 2 cases we have now without too much adjustments.

I tested the migration manually but only with my local database. I'll test it with OSD data later. Note that this basically throws out all size limits set on parent group level and instead assigns the sum of all limits within the contained parent groups. Due to the column renaming (which is actually a deletion and an insertion, at least with DBIx) I needed to spit the migration into 2 separate scripts (but it is still only one migration).

Martchus commented:
The migration ran fine locally with OSD data:

```
[info] [pid:2389] Writing default asset limits of parent groups to temporary file.
[info] [pid:2389] Default-initializing size limit for assets on parent level to sum of contained job groups.
[info] [pid:2389]  -> setting size limit of parent 26 to 40 GiB
[info] [pid:2389]  -> setting size limit of parent 1 to 1372 GiB
[info] [pid:2389]  -> setting size limit of parent 23 to 40 GiB
[info] [pid:2389]  -> setting size limit of parent 25 to 273 GiB
[info] [pid:2389]  -> setting size limit of parent 20 to 45 GiB
[info] [pid:2389]  -> setting size limit of parent 15 to 555 GiB
[info] [pid:2389]  -> setting size limit of parent 27 to 8 GiB
[info] [pid:2389]  -> setting size limit of parent 9 to 795 GiB
[info] [pid:2389]  -> setting size limit of parent 19 to 709 GiB
[info] [pid:2389]  -> setting size limit of parent 8 to 459 GiB
[info] [pid:2389]  -> setting size limit of parent 21 to 786 GiB
[info] [pid:2389]  -> setting size limit of parent 7 to 405 GiB
[info] [pid:2389]  -> setting size limit of parent 24 to 125 GiB
```

I only checked parent group `SLES 12` (ID 19) for exact correctness, though. It previously had a default of 5 GiB for its job groups and now has a parent-group limit of 709 GiB.

The explicitly set job group limits are preserved but ignored:
![screenshot_20191128_154558](https://user-images.githubusercontent.com/10248953/69815267-40805280-11f6-11ea-9b93-b211aa297856.png)

Note that this PR also increases the default limit from 100 GiB to 250 GiB because 100 GiB might be a bit less for parent groups.

okurz commented:
Now with the "throw out the old part" I am pretty much scared and lost.

> instead assigns the sum of all limits within the contained parent groups.

I don't understand that sentence. assigns to *what*? And I guess you mean "within the *containing* parent groups"? or "child groups"?!?

okurz commented:
I recommend to just keep the 100 GiB default. That is big enough for any "fresh setup".

Martchus commented:
Likely the migration code itself is easier to understand: https://github.com/os-autoinst/openQA/pull/2536/files#diff-0ab4be53e80ae97c2170bd226a502110R32

Martchus commented:
> I recommend to just keep the 100 GiB default.

I moved my change for the default out to a separate commit on a different branch. We can come back to it later if we want.

okurz commented:
Can you update the PR description with the current state? I assume you think of the PR as ready to be merged but just to make sure.

Martchus commented:
Yes, it is ready from my point of view.

okurz commented:
@Martchus https://build.opensuse.org/package/live_build_log/devel:openQA:TestGithub:PR-2536/openQA/openSUSE_Tumbleweed/x86_64 shows t/api/04-jobs.t failed. I have not seen this as a "sporadic" problem in before so we should take this serious.

@coolo I am not sure how this is supposed to work in practice. E.g. take the OSD case for SLE15. We have 10+ job groups where we can assume one team behind each. For now every now and then the teams bump up the limit in their job groups. Every some months we reduce the limit on every single job group to not run out of space. With this PR we would have 10+ teams using up the "shared quota" of a parent job group and 10+ teams would eventually bump the parent group limit from time to time. Is this really better?

okurz commented:
As discussed in jangouts meeting it should be still possible to disable quotas in parent job groups. I understood with the current state of the PR it is not possible move back easily. Maybe we can simply reuse a setting of "0" to disable parent job group quota. That should be simple. If it's not, don't care.

Martchus commented:
* Removed the previously mentioned migration.
* Added a migration to assign job group asset limits explicitly to all job groups which so far inherited the asset limit from the parent job group.
* The UI allows to clear the limit again:
    1. ![screenshot_20191210_153540](https://user-images.githubusercontent.com/10248953/70538676-c9867a80-1b62-11ea-91f8-3c069ac81619.png)
    2. ![screenshot_20191210_153643](https://user-images.githubusercontent.com/10248953/70538723-e02cd180-1b62-11ea-9a16-1e66a2c18163.png)

Martchus commented:
Ok, but because it is quite a big change maybe somebody else in the team wants to review it?

Martchus commented:
Regenerated the DB migration. This is actually a little bit work so it would be nice the PR could be merged before yet another DB migration lands on master. I haven't merged it yesterday because we had cleanup problems anyways. I also don't like to merge it so close before almost everybody is on vacation.

okurz commented:
so
1. another tidy needed
2. if you do not want it to be merged before holiday you can split it, e.g. db preparations first?

Martchus commented:
2. would be more work than rebasing the database migration once more.

kraih commented on dbicdh/_common/upgrade/84-85/000-backup-asset-size-defaults-of-parents.pl:
```diff
@@ -0,0 +1,54 @@
+
+    my $dbh          = $schema->storage->dbh;
+    my $select_query = $dbh->prepare('select id, default_size_limit_gb from job_group_parents;');
+    my $update_query
```
Further down you used `$update_sth`, that's more consistent.

kraih commented on templates/admin/group/group_property_editor.html.ep:
```diff
@@ -26,15 +26,45 @@
+                        <span id="editor-size-limit-warning">
+                            <div>
+                                The size limit for this group is overriden by its <a href="<%= url_for 'admin_edit_parent_group', groupid => $parent_group->id %>">parent job
+                                group</a> to <%= OpenQA::Utils::human_readable_size($overridden_asset_limit * 1024 * 1024 * 1024) %>. Therefore
```
You can import functions into templates (`% use OpenQA::Utils 'human_readable_size';`), no need to always use the full class name.

Martchus commented on lib/OpenQA/Schema/Result/JobGroups.pm:
```diff
@@ -122,7 +122,14 @@ sub _get_column_or_default {
+    if (defined(my $own_value = $self->get_column('size_limit_gb'))) {
+        return $own_value;
+    }
+    return $OpenQA::Utils::app->config->{default_group_limits}->{asset_size_limit};
```
This PR is getting big enough. I'd rather postpone this to a refactoring where we get rid of all those occurrences at once.
